### PR TITLE
first try

### PR DESCRIPTION
**Pull request by orm011**
The contents written to db seem reasonable, but the modified test is failing, as there is no errortype in the column.
Also, currently fails with sqlalchemy constraint violation if I set the column to nullable.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/52)

### CODE REVIEW

File: pixeltable/store.py [hunk 1]
```diff
@@ -243,6 +247,25 @@ def insert_rows(
                     table_rows: List[Dict[str, Any]] = []
                     for row_idx in range(batch_start_idx, min(batch_start_idx + batch_size, len(row_batch))):
                         row = row_batch[row_idx]
+                        for slot_idx in row.img_slot_idxs:
+                            if row.file_paths[slot_idx] is not None:
+                                try:
+                                    PIL.Image.open(row.file_paths[slot_idx])
+                                except Exception as e:
+                                    # row.vals[slot_idx] = None
+                                    row.has_val[slot_idx] = False
```
Comment (mkornacker) [inline pixeltable/store.py]: This needs to happen when the slot in the datarow gets set. That way, you can take that into account and avoid evaluating exprs on that slot.

**Review by mkornacker — CHANGES_REQUESTED** [review]

File: pixeltable/catalog/insertable_table.py [hunk 2]
```diff
@@ -113,7 +113,6 @@ def insert(self, rows: List[List[Any]], columns: List[str] = [], print_stats: bo
                 f'The number of column values in rows ({len(rows[0])}) does not match the given number of column names '
                 f'({", ".join(columns)})')
```
Comment (mkornacker) [inline pixeltable/catalog/insertable_table.py:118]: insert() now needs to take a parameter ignore_errors: bool = False

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/52)* status: 0 of 15 files reviewed, 2 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/store.py` line 256 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/52#gh-1426046333:-NltayHj7RzbgFoRbLZq:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/f93b4539666316a710457806525273d8ddc31bbc/pixeltable/store.py#L256)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

This needs to happen when the slot in the datarow gets set. That way, you can take that into account and avoid evaluating exprs on that slot.
</blockquote></details>

Done.
___
*[`pixeltable/catalog/insertable_table.py` line 115 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/52#gh-1426223041:-Nltb3deBxfIDjgkQXah:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/f93b4539666316a710457806525273d8ddc31bbc/pixeltable/catalog/insertable_table.py#L115)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

insert() now needs to take a parameter ignore_errors: bool = False
</blockquote></details>

Done.

**Review by mkornacker — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/52)* status: 0 of 16 files reviewed, 8 unresolved discussions (waiting on @orm011)

___
*[`pixeltable/exec/validate_data_node.py` line 8 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluGTzd2P50syLvlRMG:-NluGTzd2P50syLvlRMH:benh7q0) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/exec/validate_data_node.py#L8)):*
> ```Python
> import pixeltable.exprs as exprs
> 
> class ValidateDataNode(ExecNode):
> ```

Let's call this DataValidationNode, to keep with the ongoing naming scheme.
___
*[`pixeltable/exprs/row_builder.py` line 286 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluIKvSBo0Oy0GhOd5Z:-NluIKvSBo0Oy0GhOd5_:b-xkflcn) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/exprs/row_builder.py#L286)):*
> ```Python
>     
>     def validate(self, data_row: DataRow, ignore_errors: bool = False) -> None:
>         for slot_idx in data_row.img_slot_idxs + data_row.video_slot_idxs:
> ```

This doesn't seem right: you only want to validate columns that were explicitly provided by the user. The plan would know what those are.

Also, this goes beyond media columns.
___
*[`pixeltable/exprs/row_builder.py` line 292 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluHlBv3tGyUvHgjaS-:-NluHlBv3tGyUvHgjaS0:b8j1xa) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/exprs/row_builder.py#L292)):*
> ```Python
>             col_type = self.table_columns_dict[slot_idx]
>             try:
>                 # Is nullity being tested here? or have we skipped it above.
> ```

I'm not following the comment
___
*[`pixeltable/tests/test_table.py` line 134 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluJD09D-kLQybqtyU1:-NluJD09D-kLQybqtyU2:bruz23y) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/tests/test_table.py#L134)):*
> ```Python
> 
>         ## Mode 1: Validation error on bad input (default)
>         with pytest.raises(exc.DataValidationError) as exc_info:
> ```

Only pixeltable.Error should bubble up. DataFrame.collect() converts the ExprEvalError into an Error.
___
*[`pixeltable/tests/test_table.py` line 154 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluJvWCEbbFKPhMf-Wg:-NluJvWCEbbFKPhMf-Wh:bmy2ynn) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/tests/test_table.py#L154)):*
> ```Python
>         # NB(orm) found no intuitive way to express this without getting data out
>         # (see https://www.notion.so/API-pain-points-60841135411c4c67bb422dc0179dc08a )
>         assert (tbl.where((tbl.is_bad_image == False)).select(tbl.img.fileurl).show()
> ```

You should use errortype here, because that's the error indication mechanism you want to test.
___
*[`pixeltable/tests/test_table.py` line 160 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluK6yG-Vv2xzjaRMhu:-NluK6yG-Vv2xzjaRMhv:b384912) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/tests/test_table.py#L160)):*
> ```Python
>                 .to_pandas()['img_fileurl'].isnull().sum()) == 1
>         
>     def test_insert_validate_video(self, test_client: pt.Client) -> None:
> ```

Pull common logic into function and then call that.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/52)* status: 0 of 19 files reviewed, 8 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/exec/validate_data_node.py` line 8 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluGTzd2P50syLvlRMG:-NmESwuU14yTx111LI-h:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/exec/validate_data_node.py#L8)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Let's call this DataValidationNode, to keep with the ongoing naming scheme.
</blockquote></details>

Done.
___
*[`pixeltable/exprs/row_builder.py` line 286 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluIKvSBo0Oy0GhOd5Z:-NmET-NfDc97msuXljby:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/exprs/row_builder.py#L286)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

This doesn't seem right: you only want to validate columns that were explicitly provided by the user. The plan would know what those are.

Also, this goes beyond media columns.
</blockquote></details>

Done.
___
*[`pixeltable/exprs/row_builder.py` line 292 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluHlBv3tGyUvHgjaS-:-NmET08S32F_rQQdYJf7:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/exprs/row_builder.py#L292)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

I'm not following the comment
</blockquote></details>

Done.
___
*[`pixeltable/tests/test_table.py` line 134 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluJD09D-kLQybqtyU1:-NmET1oo0jjM--eUu8Y1:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/tests/test_table.py#L134)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Only pixeltable.Error should bubble up. DataFrame.collect() converts the ExprEvalError into an Error.
</blockquote></details>

Done.
___
*[`pixeltable/tests/test_table.py` line 154 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluJvWCEbbFKPhMf-Wg:-NmET8wg2Db2FNHVke-T:b-o6mahr) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/tests/test_table.py#L154)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

You should use errortype here, because that's the error indication mechanism you want to test.
</blockquote></details>

I'm just testing url is not set.Error type was set earlier.
___
*[`pixeltable/tests/test_table.py` line 160 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/52#-NluK6yG-Vv2xzjaRMhu:-NmETDLd3HhYVCAvXdzn:bnwixep) ([raw file](https://github.com/mkornacker/pixeltable/blob/3bfc44cca7eb6273a96b4ec2100935f7d74efc42/pixeltable/tests/test_table.py#L160)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Pull common logic into function and then call that.
</blockquote></details>

Done. re-used stuff for audio as well.

**Comment by orm011** [comment]
was merged manually: commit 43b6f43.

**Pull request closed without merging**